File: packages/neko-engine/packages/effects-core/src/algorithms/easing.ts
/**
 * Easing Algorithms
 *
 * Pure easing function implementations.
 * Platform-agnostic, can be used in JS, WGSL, and Rust.
 */

import type { EasingType, EasingFunction, CubicBezierEasing, SpringConfig, SpringState } from '../types/easing';

// =============================================================================
// Standard Easing Functions
// =============================================================================

// Linear
export const linear: EasingFunction = (t) => t;

// Quad
export const easeInQuad: EasingFunction = (t) => t * t;
export const easeOutQuad: EasingFunction = (t) => t * (2 - t);
export const easeInOutQuad: EasingFunction = (t) =>
	t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t;

// Cubic
export const easeInCubic: EasingFunction = (t) => t * t * t;
export const easeOutCubic: EasingFunction = (t) => (--t) * t * t + 1;
export const easeInOutCubic: EasingFunction = (t) =>
	t < 0.5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1;

// Quart
export const easeInQuart: EasingFunction = (t) => t * t * t * t;
export const easeOutQuart: EasingFunction = (t) => 1 - (--t) * t * t * t;
export const easeInOutQuart: EasingFunction = (t) =>
	t < 0.5 ? 8 * t * t * t * t : 1 - 8 * (--t) * t * t * t;

// Quint
export const easeInQuint: EasingFunction = (t) => t * t * t * t * t;
export const easeOutQuint: EasingFunction = (t) => 1 + (--t) * t * t * t * t;
export const easeInOutQuint: EasingFunction = (t) =>
	t < 0.5 ? 16 * t * t * t * t * t : 1 + 16 * (--t) * t * t * t * t;

// Sine
export const easeInSine: EasingFunction = (t) => 1 - Math.cos((t * Math.PI) / 2);
export const easeOutSine: EasingFunction = (t) => Math.sin((t * Math.PI) / 2);
export const easeInOutSine: EasingFunction = (t) => -(Math.cos(Math.PI * t) - 1) / 2;

// Expo
export const easeInExpo: EasingFunction = (t) =>
	t === 0 ? 0 : Math.pow(2, 10 * t - 10);
export const easeOutExpo: EasingFunction = (t) =>
	t === 1 ? 1 : 1 - Math.pow(2, -10 * t);
export const easeInOutExpo: EasingFunction = (t) =>
	t === 0 ? 0 : t === 1 ? 1 : t < 0.5
		? Math.pow(2, 20 * t - 10) / 2
		: (2 - Math.pow(2, -20 * t + 10)) / 2;

// Circ
export const easeInCirc: EasingFunction = (t) => 1 - Math.sqrt(1 - t * t);
export const easeOutCirc: EasingFunction = (t) => Math.sqrt(1 - (--t) * t);
export const easeInOutCirc: EasingFunction = (t) =>
	t < 0.5
		? (1 - Math.sqrt(1 - 4 * t * t)) / 2
		: (Math.sqrt(1 - Math.pow(-2 * t + 2, 2)) + 1) / 2;

// Back
const c1 = 1.70158;
const c2 = c1 * 1.525;
const c3 = c1 + 1;

export const easeInBack: EasingFunction = (t) => c3 * t * t * t - c1 * t * t;
export const easeOutBack: EasingFunction = (t) =>
	1 + c3 * Math.pow(t - 1, 3) + c1 * Math.pow(t - 1, 2);
export const easeInOutBack: EasingFunction = (t) =>
	t < 0.5
		? (Math.pow(2 * t, 2) * ((c2 + 1) * 2 * t - c2)) / 2
		: (Math.pow(2 * t - 2, 2) * ((c2 + 1) * (t * 2 - 2) + c2) + 2) / 2;

// Elastic
const c4 = (2 * Math.PI) / 3;
const c5 = (2 * Math.PI) / 4.5;

export const easeInElastic: EasingFunction = (t) =>
	t === 0 ? 0 : t === 1 ? 1 : -Math.pow(2, 10 * t - 10) * Math.sin((t * 10 - 10.75) * c4);
export const easeOutElastic: EasingFunction = (t) =>
	t === 0 ? 0 : t === 1 ? 1 : Math.pow(2, -10 * t) * Math.sin((t * 10 - 0.75) * c4) + 1;
export const easeInOutElastic: EasingFunction = (t) =>
	t === 0 ? 0 : t === 1 ? 1 : t < 0.5
		? -(Math.pow(2, 20 * t - 10) * Math.sin((20 * t - 11.125) * c5)) / 2
		: (Math.pow(2, -20 * t + 10) * Math.sin((20 * t - 11.125) * c5)) / 2 + 1;

// Bounce
export const easeOutBounce: EasingFunction = (t) => {
	const n1 = 7.5625;
	const d1 = 2.75;

	if (t < 1 / d1) {
		return n1 * t * t;
	} else if (t < 2 / d1) {
		return n1 * (t -= 1.5 / d1) * t + 0.75;
	} else if (t < 2.5 / d1) {
		return n1 * (t -= 2.25 / d1) * t + 0.9375;
	} else {
		return n1 * (t -= 2.625 / d1) * t + 0.984375;
	}
};

export const easeInBounce: EasingFunction = (t) => 1 - easeOutBounce(1 - t);
export const easeInOutBounce: EasingFunction = (t) =>
	t < 0.5
		? (1 - easeOutBounce(1 - 2 * t)) / 2
		: (1 + easeOutBounce(2 * t - 1)) / 2;

// =============================================================================
// Easing Function Registry
// =============================================================================

const EASING_FUNCTIONS: Record<Exclude<EasingType, 'cubicBezier'>, EasingFunction> = {
	linear,
	easeInQuad,
	easeOutQuad,
	easeInOutQuad,
	easeInCubic,
	easeOutCubic,
	easeInOutCubic,
	easeInQuart,
	easeOutQuart,
	easeInOutQuart,
	easeInQuint,
	easeOutQuint,
	easeInOutQuint,
	easeInSine,
	easeOutSine,
	easeInOutSine,
	easeInExpo,
	easeOutExpo,
	easeInOutExpo,
	easeInCirc,
	easeOutCirc,
	easeInOutCirc,
	easeInBack,
	easeOutBack,
	easeInOutBack,
	easeInElastic,
	easeOutElastic,
	easeInOutElastic,
	easeInBounce,
	easeOutBounce,
	easeInOutBounce,
};

/**
 * Get easing function by type
 */
export function getEasingFunction(type: EasingType): EasingFunction {
	if (type === 'cubicBezier') {
		throw new Error('Use createCubicBezierEasing for custom bezier curves');
	}
	return EASING_FUNCTIONS[type];
}

// =============================================================================
// Cubic Bezier
// =============================================================================

/**
 * Create a cubic bezier easing function
 * Based on WebKit's implementation
 */
export function createCubicBezierEasing(config: CubicBezierEasing): EasingFunction {
	const { x1, y1, x2, y2 } = config;

	// Special cases
	if (x1 === y1 && x2 === y2) {
		return linear;
	}

	// Calculate coefficients
	const cx = 3 * x1;
	const bx = 3 * (x2 - x1) - cx;
	const ax = 1 - cx - bx;

	const cy = 3 * y1;
	const by = 3 * (y2 - y1) - cy;
	const ay = 1 - cy - by;

	const sampleCurveX = (t: number) => ((ax * t + bx) * t + cx) * t;
	const sampleCurveY = (t: number) => ((ay * t + by) * t + cy) * t;
	const sampleCurveDerivativeX = (t: number) => (3 * ax * t + 2 * bx) * t + cx;

	// Newton-Raphson iteration to find t for given x
	const solveCurveX = (x: number): number => {
		let t = x;
		for (let i = 0; i < 8; i++) {
			const currentX = sampleCurveX(t) - x;
			if (Math.abs(currentX) < 1e-6) {
				return t;
			}
			const derivative = sampleCurveDerivativeX(t);
			if (Math.abs(derivative) < 1e-6) {
				break;
			}
			t -= currentX / derivative;
		}

		// Fall back to bisection
		let t0 = 0;
		let t1 = 1;
		t = x;

		while (t0 < t1) {
			const currentX = sampleCurveX(t);
			if (Math.abs(currentX - x) < 1e-6) {
				return t;
			}
			if (x > currentX) {
				t0 = t;
			} else {
				t1 = t;
			}
			t = (t1 - t0) * 0.5 + t0;
		}

		return t;
	};

	return (x: number) => sampleCurveY(solveCurveX(x));
}

// =============================================================================
// Spring Physics
// =============================================================================

/**
 * Default spring configuration
 */
export const DEFAULT_SPRING_CONFIG: SpringConfig = {
	mass: 1,
	stiffness: 100,
	damping: 10,
	velocity: 0,
};

/**
 * Calculate spring physics step
 * @param state Current spring state
 * @param target Target position
 * @param config Spring configuration
 * @param dt Delta time in seconds
 * @returns New spring state
 */
export function springStep(
	state: SpringState,
	target: number,
	config: SpringConfig,
	dt: number
): SpringState {
	const { mass, stiffness, damping } = config;
	const { position, velocity } = state;

	// Spring force: F = -k * x
	const springForce = -stiffness * (position - target);
	// Damping force: F = -c * v
	const dampingForce = -damping * velocity;
	// Total force
	const force = springForce + dampingForce;
	// Acceleration: a = F / m
	const acceleration = force / mass;

	// Euler integration
	const newVelocity = velocity + acceleration * dt;
	const newPosition = position + newVelocity * dt;

	return {
		position: newPosition,
		velocity: newVelocity,
	};
}

/**
 * Check if spring has settled
 */
export function isSpringSettled(
	state: SpringState,
	target: number,
	threshold = 0.001
): boolean {
	return (
		Math.abs(state.position - target) < threshold &&
		Math.abs(state.velocity) < threshold
	);
}

// =============================================================================
// Interpolation Helpers
// =============================================================================

/**
 * Linear interpolation
 */
export function lerp(a: number, b: number, t: number): number {
	return a + (b - a) * t;
}

/**
 * Clamp value between min and max
 */
export function clamp(value: number, min: number, max: number): number {
	return Math.min(Math.max(value, min), max);
}

/**
 * Map value from one range to another
 */
export function mapRange(
	value: number,
	inMin: number,
	inMax: number,
	outMin: number,
	outMax: number
): number {
	return ((value - inMin) / (inMax - inMin)) * (outMax - outMin) + outMin;
}

/**
 * Smooth step interpolation
 */
export function smoothstep(edge0: number, edge1: number, x: number): number {
	const t = clamp((x - edge0) / (edge1 - edge0), 0, 1);
	return t * t * (3 - 2 * t);
}

/**
 * Smoother step interpolation (Ken Perlin's version)
 */
export function smootherstep(edge0: number, edge1: number, x: number): number {
	const t = clamp((x - edge0) / (edge1 - edge0), 0, 1);
	return t * t * t * (t * (t * 6 - 15) + 10);
}

File: packages/neko-engine/packages/media-processor-rs/src/animation/easing.rs
//! Easing Functions
//!
//! Implementation of standard easing curves for animations.
//! Compatible with CSS easing and After Effects style curves.

use std::f64::consts::PI;

/// Easing function types
#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub enum EasingType {
    // Linear
    #[default]
    Linear,
    // Quad
    EaseInQuad,
    EaseOutQuad,
    EaseInOutQuad,
    // Cubic
    EaseInCubic,
    EaseOutCubic,
    EaseInOutCubic,
    // Quart
    EaseInQuart,
    EaseOutQuart,
    EaseInOutQuart,
    // Quint
    EaseInQuint,
    EaseOutQuint,
    EaseInOutQuint,
    // Sine
    EaseInSine,
    EaseOutSine,
    EaseInOutSine,
    // Expo
    EaseInExpo,
    EaseOutExpo,
    EaseInOutExpo,
    // Circ
    EaseInCirc,
    EaseOutCirc,
    EaseInOutCirc,
    // Back
    EaseInBack,
    EaseOutBack,
    EaseInOutBack,
    // Elastic
    EaseInElastic,
    EaseOutElastic,
    EaseInOutElastic,
    // Bounce
    EaseInBounce,
    EaseOutBounce,
    EaseInOutBounce,
    // Custom cubic bezier
    CubicBezier(f64, f64, f64, f64), // x1, y1, x2, y2
}

impl EasingType {
    /// Parse easing type from string
    pub fn from_str(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "linear" => EasingType::Linear,
            "easeinquad" | "ease_in_quad" => EasingType::EaseInQuad,
            "easeoutquad" | "ease_out_quad" => EasingType::EaseOutQuad,
            "easeinoutquad" | "ease_in_out_quad" => EasingType::EaseInOutQuad,
            "easeincubic" | "ease_in_cubic" => EasingType::EaseInCubic,
            "easeoutcubic" | "ease_out_cubic" => EasingType::EaseOutCubic,
            "easeinoutcubic" | "ease_in_out_cubic" => EasingType::EaseInOutCubic,
            "easeinquart" | "ease_in_quart" => EasingType::EaseInQuart,
            "easeoutquart" | "ease_out_quart" => EasingType::EaseOutQuart,
            "easeinoutquart" | "ease_in_out_quart" => EasingType::EaseInOutQuart,
            "easeinquint" | "ease_in_quint" => EasingType::EaseInQuint,
            "easeoutquint" | "ease_out_quint" => EasingType::EaseOutQuint,
            "easeinoutquint" | "ease_in_out_quint" => EasingType::EaseInOutQuint,
            "easeinsine" | "ease_in_sine" => EasingType::EaseInSine,
            "easeoutsine" | "ease_out_sine" => EasingType::EaseOutSine,
            "easeinoutsine" | "ease_in_out_sine" => EasingType::EaseInOutSine,
            "easeinexpo" | "ease_in_expo" => EasingType::EaseInExpo,
            "easeoutexpo" | "ease_out_expo" => EasingType::EaseOutExpo,
            "easeinoutexpo" | "ease_in_out_expo" => EasingType::EaseInOutExpo,
            "easeincirc" | "ease_in_circ" => EasingType::EaseInCirc,
            "easeoutcirc" | "ease_out_circ" => EasingType::EaseOutCirc,
            "easeinoutcirc" | "ease_in_out_circ" => EasingType::EaseInOutCirc,
            "easeinback" | "ease_in_back" => EasingType::EaseInBack,
            "easeoutback" | "ease_out_back" => EasingType::EaseOutBack,
            "easeinoutback" | "ease_in_out_back" => EasingType::EaseInOutBack,
            "easeinelastic" | "ease_in_elastic" => EasingType::EaseInElastic,
            "easeoutelastic" | "ease_out_elastic" => EasingType::EaseOutElastic,
            "easeinoutelastic" | "ease_in_out_elastic" => EasingType::EaseInOutElastic,
            "easeinbounce" | "ease_in_bounce" => EasingType::EaseInBounce,
            "easeoutbounce" | "ease_out_bounce" => EasingType::EaseOutBounce,
            "easeinoutbounce" | "ease_in_out_bounce" => EasingType::EaseInOutBounce,
            _ => EasingType::Linear,
        }
    }
}

/// Easing function evaluation
pub struct Easing;

impl Easing {
    /// Evaluate easing function at time t (0.0 - 1.0)
    pub fn evaluate(easing: EasingType, t: f64) -> f64 {
        let t = t.clamp(0.0, 1.0);

        match easing {
            EasingType::Linear => t,

            // Quad
            EasingType::EaseInQuad => t * t,
            EasingType::EaseOutQuad => t * (2.0 - t),
            EasingType::EaseInOutQuad => {
                if t < 0.5 {
                    2.0 * t * t
                } else {
                    -1.0 + (4.0 - 2.0 * t) * t
                }
            }

            // Cubic
            EasingType::EaseInCubic => t * t * t,
            EasingType::EaseOutCubic => {
                let t1 = t - 1.0;
                t1 * t1 * t1 + 1.0
            }
            EasingType::EaseInOutCubic => {
                if t < 0.5 {
                    4.0 * t * t * t
                } else {
                    let t1 = 2.0 * t - 2.0;
                    (t1 * t1 * t1 + 2.0) / 2.0
                }
            }

            // Quart
            EasingType::EaseInQuart => t * t * t * t,
            EasingType::EaseOutQuart => {
                let t1 = t - 1.0;
                1.0 - t1 * t1 * t1 * t1
            }
            EasingType::EaseInOutQuart => {
                if t < 0.5 {
                    8.0 * t * t * t * t
                } else {
                    let t1 = t - 1.0;
                    1.0 - 8.0 * t1 * t1 * t1 * t1
                }
            }

            // Quint
            EasingType::EaseInQuint => t * t * t * t * t,
            EasingType::EaseOutQuint => {
                let t1 = t - 1.0;
                t1 * t1 * t1 * t1 * t1 + 1.0
            }
            EasingType::EaseInOutQuint => {
                if t < 0.5 {
                    16.0 * t * t * t * t * t
                } else {
                    let t1 = 2.0 * t - 2.0;
                    (t1 * t1 * t1 * t1 * t1 + 2.0) / 2.0
                }
            }

            // Sine
            EasingType::EaseInSine => 1.0 - (t * PI / 2.0).cos(),
            EasingType::EaseOutSine => (t * PI / 2.0).sin(),
            EasingType::EaseInOutSine => -(PI * t).cos() / 2.0 + 0.5,

            // Expo
            EasingType::EaseInExpo => {
                if t == 0.0 {
                    0.0
                } else {
                    (2.0_f64).powf(10.0 * (t - 1.0))
                }
            }
            EasingType::EaseOutExpo => {
                if t == 1.0 {
                    1.0
                } else {
                    1.0 - (2.0_f64).powf(-10.0 * t)
                }
            }
            EasingType::EaseInOutExpo => {
                if t == 0.0 {
                    0.0
                } else if t == 1.0 {
                    1.0
                } else if t < 0.5 {
                    (2.0_f64).powf(20.0 * t - 10.0) / 2.0
                } else {
                    (2.0 - (2.0_f64).powf(-20.0 * t + 10.0)) / 2.0
                }
            }

            // Circ
            EasingType::EaseInCirc => 1.0 - (1.0 - t * t).sqrt(),
            EasingType::EaseOutCirc => (1.0 - (t - 1.0) * (t - 1.0)).sqrt(),
            EasingType::EaseInOutCirc => {
                if t < 0.5 {
                    (1.0 - (1.0 - 4.0 * t * t).sqrt()) / 2.0
                } else {
                    let t1 = -2.0 * t + 2.0;
                    ((1.0 - t1 * t1).sqrt() + 1.0) / 2.0
                }
            }

            // Back
            EasingType::EaseInBack => {
                let c1 = 1.70158;
                let c3 = c1 + 1.0;
                c3 * t * t * t - c1 * t * t
            }
            EasingType::EaseOutBack => {
                let c1 = 1.70158;
                let c3 = c1 + 1.0;
                let t1 = t - 1.0;
                1.0 + c3 * t1 * t1 * t1 + c1 * t1 * t1
            }
            EasingType::EaseInOutBack => {
                let c1 = 1.70158;
                let c2 = c1 * 1.525;
                if t < 0.5 {
                    (4.0 * t * t * ((c2 + 1.0) * 2.0 * t - c2)) / 2.0
                } else {
                    let t1 = 2.0 * t - 2.0;
                    (t1 * t1 * ((c2 + 1.0) * t1 + c2) + 2.0) / 2.0
                }
            }

            // Elastic
            EasingType::EaseInElastic => {
                let c4 = 2.0 * PI / 3.0;
                if t == 0.0 {
                    0.0
                } else if t == 1.0 {
                    1.0
                } else {
                    -(2.0_f64).powf(10.0 * t - 10.0) * ((t * 10.0 - 10.75) * c4).sin()
                }
            }
            EasingType::EaseOutElastic => {
                let c4 = 2.0 * PI / 3.0;
                if t == 0.0 {
                    0.0
                } else if t == 1.0 {
                    1.0
                } else {
                    (2.0_f64).powf(-10.0 * t) * ((t * 10.0 - 0.75) * c4).sin() + 1.0
                }
            }
            EasingType::EaseInOutElastic => {
                let c5 = 2.0 * PI / 4.5;
                if t == 0.0 {
                    0.0
                } else if t == 1.0 {
                    1.0
                } else if t < 0.5 {
                    -(2.0_f64).powf(20.0 * t - 10.0) * ((20.0 * t - 11.125) * c5).sin() / 2.0
                } else {
                    (2.0_f64).powf(-20.0 * t + 10.0) * ((20.0 * t - 11.125) * c5).sin() / 2.0 + 1.0
                }
            }

            // Bounce
            EasingType::EaseInBounce => 1.0 - Self::bounce_out(1.0 - t),
            EasingType::EaseOutBounce => Self::bounce_out(t),
            EasingType::EaseInOutBounce => {
                if t < 0.5 {
                    (1.0 - Self::bounce_out(1.0 - 2.0 * t)) / 2.0
                } else {
                    (1.0 + Self::bounce_out(2.0 * t - 1.0)) / 2.0
                }
            }

            // Cubic Bezier
            EasingType::CubicBezier(x1, y1, x2, y2) => {
                Self::cubic_bezier(t, x1, y1, x2, y2)
            }
        }
    }

    /// Bounce easing out helper
    fn bounce_out(t: f64) -> f64 {
        let n1 = 7.5625;
        let d1 = 2.75;

        if t < 1.0 / d1 {
            n1 * t * t
        } else if t < 2.0 / d1 {
            let t = t - 1.5 / d1;
            n1 * t * t + 0.75
        } else if t < 2.5 / d1 {
            let t = t - 2.25 / d1;
            n1 * t * t + 0.9375
        } else {
            let t = t - 2.625 / d1;
            n1 * t * t + 0.984375
        }
    }

    /// Cubic bezier easing
    fn cubic_bezier(t: f64, x1: f64, y1: f64, x2: f64, y2: f64) -> f64 {
        // Newton-Raphson iteration to find t for x
        let mut t_approx = t;
        for _ in 0..8 {
            let x = Self::bezier_calc(t_approx, x1, x2) - t;
            if x.abs() < 1e-6 {
                break;
            }
            let dx = Self::bezier_derivative(t_approx, x1, x2);
            if dx.abs() < 1e-6 {
                break;
            }
            t_approx -= x / dx;
        }

        Self::bezier_calc(t_approx, y1, y2)
    }

    /// Calculate bezier value at t
    fn bezier_calc(t: f64, p1: f64, p2: f64) -> f64 {
        let t2 = t * t;
        let t3 = t2 * t;
        let mt = 1.0 - t;
        let mt2 = mt * mt;

        3.0 * mt2 * t * p1 + 3.0 * mt * t2 * p2 + t3
    }

    /// Calculate bezier derivative at t
    fn bezier_derivative(t: f64, p1: f64, p2: f64) -> f64 {
        let t2 = t * t;
        let mt = 1.0 - t;

        3.0 * mt * mt * p1 + 6.0 * mt * t * (p2 - p1) + 3.0 * t2 * (1.0 - p2)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_linear() {
        assert!((Easing::evaluate(EasingType::Linear, 0.0) - 0.0).abs() < 1e-6);
        assert!((Easing::evaluate(EasingType::Linear, 0.5) - 0.5).abs() < 1e-6);
        assert!((Easing::evaluate(EasingType::Linear, 1.0) - 1.0).abs() < 1e-6);
    }

    #[test]
    fn test_ease_in_quad() {
        assert!((Easing::evaluate(EasingType::EaseInQuad, 0.0) - 0.0).abs() < 1e-6);
        assert!((Easing::evaluate(EasingType::EaseInQuad, 0.5) - 0.25).abs() < 1e-6);
        assert!((Easing::evaluate(EasingType::EaseInQuad, 1.0) - 1.0).abs() < 1e-6);
    }

    #[test]
    fn test_ease_out_quad() {
        assert!((Easing::evaluate(EasingType::EaseOutQuad, 0.0) - 0.0).abs() < 1e-6);
        assert!((Easing::evaluate(EasingType::EaseOutQuad, 0.5) - 0.75).abs() < 1e-6);
        assert!((Easing::evaluate(EasingType::EaseOutQuad, 1.0) - 1.0).abs() < 1e-6);
    }

    #[test]
    fn test_easing_from_str() {
        assert_eq!(EasingType::from_str("linear"), EasingType::Linear);
        assert_eq!(EasingType::from_str("easeInQuad"), EasingType::EaseInQuad);
        assert_eq!(EasingType::from_str("ease_out_cubic"), EasingType::EaseOutCubic);
        assert_eq!(EasingType::from_str("unknown"), EasingType::Linear);
    }

    #[test]
    fn test_cubic_bezier() {
        // CSS ease: cubic-bezier(0.25, 0.1, 0.25, 1)
        let result = Easing::evaluate(EasingType::CubicBezier(0.25, 0.1, 0.25, 1.0), 0.5);
        assert!(result > 0.5, "CSS ease should be faster than linear at 0.5");
    }

    #[test]
    fn test_bounce() {
        let result = Easing::evaluate(EasingType::EaseOutBounce, 0.9);
        // Should bounce near the end
        assert!(result > 0.9);
    }
}
